xla/README.md

```diff
@@ -0,0 +1,124 @@
+
+</details>
+
+Once inside the machine, clone the repository and install the dependencies:
+
+```shell
+git clone https://github.com/Lightning-AI/lit-gpt
+cd lit-gpt
+pip install -r requirements.txt
+```
+
+Install Optimized BLAS:
+
+```shell
+sudo apt update
+sudo apt install libopenblas-dev
+```
+
+Since Lit-GPT requires a torch version newer than torch 2.0.0, manually install nightly builds of torch and torch_xla:
+
+```shell
+pip install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-nightly-cp38-cp38-linux_x86_64.whl
+pip install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-nightly-cp38-cp38-linux_x86_64.whl
+```
+
+While computations will run by default using the new PjRT runtime, it is recommended to set the following environment variables:
+
+```shell
+export ALLOW_MULTIPLE_LIBTPU_LOAD=1
+export PJRT_DEVICE=TPU
+```
+
+> [!NOTE]\
+> An extensive guide on setup and available options can be found [here](https://cloud.google.com/tpu/docs/v4-users-guide).
+
+Since a new machine was created, you may need to download pretrained weights.
+They can be copied to the machine using `gcloud compute tpus tpu-vm scp`, or you can follow the steps described in our [downloading guide](download_stablelm.md).
+
+It is also recommended to set up a persistent disk from which to load checkpoints.
+Follow [this guide](https://cloud.google.com/tpu/docs/setup-persistent-disk#setting_up_a_tpu_vm_and_a_persistent_disk) to do so.
+Read-write disks are not supported in multihost VM setups, so persistent disks cannot be used to save checkpoints in that case.
+Persistent disks can still be useful in read-only mode to load pretrained weights before finetuning or inference.
+In multihost settings, FSDP will save checkpoint shards per host and consolidate them into a single checkpoint.
+For safekeeping, it is recommended to upload the consolidated checkpoints to a Google Cloud bucket.
+Alternatively, you can use the `scp` command to transfer these checkpoints from the TPU VM periodically, although this is not implemented in our scripts.
+
+
+## Inference
+
+This project provides custom versions of the regular recipes to run with XLA in the `xla` directory.
+To generate text, use the following command:
+
+```shell
+python3 xla/generate/base.py --prompt "Hello, my name is" --num_samples 3
+```
+
+For the first generation, this command will take around 17 seconds as XLA needs to compile the graph.
+Subsequent generations will take around 2 seconds.
+
+## Fine-tuning
+
+To get started fine-tuning Falcon 7B with adapter, run the following command:
+
+```shell
+python3 xla/finetune/adapter.py --checkpoint_dir checkpoints/tiiuae/falcon-7b --precision bf16-true
+```
+
+To generate text with the adapter fine-tuned model weights, use the following command:
+
+```shell
+python3 xla/generate/adapter.py --checkpoint_dir checkpoints/tiiuae/falcon-7b --precision bf16-true --adapter_path out/adapter/alpaca/lit_model_adapter_finetuned.pth --devices 4
```

Comment:
since we are using `xla/generate/adapter.py`, do we need to still specify `devices` as a command line arg?

Reply:
Good point